### PR TITLE
feat: non-interactive CLI mode for AI agents and CI

### PR DESCRIPTION
**Pull request by fernandomg**
## Summary

Adds a non-interactive mode so AI agents and CI pipelines can install dAppBooster projects via flags instead of the interactive TUI.

## Changes

- `--name`, `--mode`, `--features`, `--ni`, `--info` flags via meow
- Non-interactive path: validates flags, runs operations, outputs JSON to stdout
- `--info` flag for agent feature discovery (structured JSON metadata)
- Operations layer (`source/operations/`) extracted from TUI components — shared by both paths
- `execFile` (no shell) for git/pnpm commands; `node:fs/promises` for async filesystem operations (sync `node:fs` for `patchPackageJson` and `existsSync` checks)
- `ink-spawn` replaced by operations layer
- 124 vitest tests focused on the agentic interface
- `AGENTS.md`, `architecture.md`, PR template

## Review feedback addressed

- `createEnvFile.ts` — replaced shell `exec('cp ...')` with `fs/promises.copyFile` (73d35d3)
- `CloneRepo.tsx`, `Install.tsx`, `FileCleanup.tsx` — catch `unknown` error type in handlers (b4eea77)
- `exec.ts` — replaced buffered `exec`/`execFile` with `spawn` to eliminate maxBuffer limit (976b2b2)
- `cli.tsx` — aligned help text bin name with `package.json` (`dappbooster`, not `dappbooster-starter`) (7797726)
- `architecture.md` — updated exec.ts docs to match spawn-based implementation (5322d74)
- `nonInteractive.ts` — deduplicate `--features` flag values preserving order (cc34435)
- `cleanupFiles.ts` — replaced shell `exec()` with `execFile()` for rm/mkdir/cp commands that don't need shell features (52d4361)
- `exec.ts` — include signal name in error when process is killed instead of misleading "exit code null" (461ea7a)
- `architecture.md` — removed incorrect "--help text" claim from `featureDefinitions.description` comment (bcb2d44)
- `CloneRepo.tsx`, `Install.tsx`, `FileCleanup.tsx` — restored per-step progress display with `onProgress` callbacks matching the original ink-spawn stepper UX (b1870e9)
- `nonInteractive.ts`, `cli.tsx` — eliminated all `process.exit()` calls, replaced with `process.exitCode` + throw/branching to prevent stdout truncation when piped (106c6f0, c8dc7fb)
- `cleanupFiles.ts` — added `-f` flag to all single-file `rm` calls for idempotent cleanup (d0e4592)
- `CloneRepo.tsx`, `Install.tsx`, `FileCleanup.tsx` — show failed step with red cross marker on error instead of hiding it (687f8d5)
- Tests — added missing `projectDirectoryExists` coverage for custom mode, fixed mock return types (a2b7c92)
- `nonInteractive.ts`, `cli.tsx` — added explicit `: unknown` on catch params for consistency (f471b2f)
- `architecture.md`, `AGENTS.md` — corrected doc-code inconsistencies: absolute claims, validation steps, trailing commas, semicolons, mocking pattern, security section (bf6067d)
- `cleanupFiles.ts`, `cloneRepo.ts` — replaced all async filesystem shell commands (rm, mkdir, cp) with `node:fs/promises` APIs, eliminating 15 child process spawns (04eaaa0)
- `Install.tsx` — fixed empty title fallback: `??` doesn't catch empty string, changed ternary default to `'Full'` directly (4f33846)
- `architecture.md` — corrected postInstall comment (agents only, not TUI), expanded How to Add a Feature guide with PostInstall.tsx and cli.tsx steps (1fd205b)

## Acceptance criteria

- [ ] `--ni --name my_app --mode full` installs everything and outputs success JSON
- [ ] `--ni --name my_app --mode custom --features demo,subgraph` installs selected features only
- [ ] `--info` outputs parseable feature metadata
- [ ] Invalid flags produce JSON errors with exit code 1
- [ ] Interactive TUI displays per-step progress with checkmarks and error markers
- [ ] All 124 tests pass, build and lint clean

## Test plan

- `pnpm build && pnpm lint && pnpm test` — all green
- `pnpm test:coverage` — nonInteractive 95%, info/config/operations 100%
- Manual verification of `--info` output and edge cases (empty features, trailing commas, duplicates, shell injection)

## Breaking changes

None.

## Checklist

- [x] Self-reviewed my own diff
- [x] Tests added or updated
- [x] Docs updated (if applicable)
- [x] No unrelated changes bundled in

### CODE REVIEW

**Comment by fernandomg** [comment]
**createEnvFile.ts** — Fixed in 73d35d3. Replaced shell `exec('cp ...')` with `fs/promises.copyFile`. No child process needed at all.

**Comment by fernandomg** [comment]
**CloneRepo.tsx, Install.tsx, FileCleanup.tsx** — Fixed in b4eea77. All three component error handlers now catch `unknown` and derive a safe message via `error instanceof Error ? error.message : String(error)`.

**Comment by fernandomg** [comment]
**exec.ts maxBuffer** — Fixed in 976b2b2. Replaced `child_process.exec/execFile` (which buffer stdout in memory) with `spawn` using `stdio: ['ignore', 'ignore', 'pipe']`. Stdout is discarded (no caller uses it), stderr is piped for error diagnostics. No buffer limit at all.

File: source/cli.tsx [hunk 1]
```diff
@@ -1,6 +1,103 @@
...
+import { getInfoOutput } from './info.js'
+import { runNonInteractive } from './nonInteractive.js'
 
-console.clear()
-render(<App />)
+const cli = meow(
+  `
+  Usage
+    $ dappbooster-starter [options]
+
```
[the remark this answers is not in the document]
Comment (fernandomg) [reply source/cli.tsx:11]: Fixed in 7797726. Updated all help text references from `dappbooster-starter` to `dappbooster` to match the actual bin name from `package.json`.

File: architecture.md [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+| `cleanupFiles(projectFolder, mode, features)` | Remove files/folders for deselected features, patch package.json scripts, remove .install-files |
+
+### Shell Execution (`source/operations/exec.ts`)
+
+Two helpers with different security profiles:
+
+- **`execFile(file, args, options)`** — uses `child_process.execFile` (no shell). Arguments are passed as an array, so user input cannot be interpreted as shell metacharacters. Use this whenever user-provided values (e.g., `projectName`) appear in the command.
+- **`exec(command, options)`** — uses `child_process.exec` (spawns a shell). Only for commands that require shell features like `$(...)` substitution. Never interpolate user input into the command string.
+
+Both helpers capture stdout/stderr (no leaking to parent process) and throw on failure with the stderr message.
```
[the remark this answers is not in the document]
Comment (fernandomg) [reply architecture.md]: Fixed in 5322d74. Updated the Shell Execution section to accurately describe the spawn-based implementation: `spawn` without shell for `execFile`, `/bin/sh -c` via `spawn` for `exec`, stdout ignored, stderr piped.

File: source/nonInteractive.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+function parseFeatures(featuresFlag: string | undefined): FeatureName[] {
+  if (!featuresFlag) {
+    return []
+  }
+
+  return featuresFlag
+    .split(',')
+    .map((f) => f.trim())
+    .filter((f) => f !== '') as FeatureName[]
```
[the remark this answers is not in the document]
Comment (fernandomg) [reply source/nonInteractive.ts]: Fixed in cc34435. `parseFeatures` now deduplicates using a `Set` while preserving insertion order. Added a test covering `demo,demo,subgraph,demo` → `['demo', 'subgraph']`.

File: source/operations/createEnvFile.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+import { exec } from './exec.js'
+
+export async function createEnvFile(projectFolder: string): Promise<void> {
+  await exec('cp .env.example .env.local', { cwd: projectFolder })
```
[the remark this answers is not in the document]
Comment (fernandomg) [reply source/operations/createEnvFile.ts]: ping!

[the remark this answers is not in the document]
Comment (fernandomg) [reply source/operations/createEnvFile.ts]: pong!

File: source/operations/cleanupFiles.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+  writeFileSync(packageJsonPath, `${JSON.stringify(packageJson, null, 2)}\n`)
+}
+
+async function cleanupDemo(projectFolder: string): Promise<void> {
+  await exec('rm -rf src/components/pageComponents/home', { cwd: projectFolder })
+  await exec('mkdir -p src/components/pageComponents/home', { cwd: projectFolder })
+  await exec('cp .install-files/home/index.tsx src/components/pageComponents/home/', {
+    cwd: projectFolder,
+  })
```
[the remark this answers is not in the document]
Comment (fernandomg) [reply source/operations/cleanupFiles.ts]: Fixed in 52d4361. Replaced all `exec()` calls in `cleanupFiles.ts` with `execFile()` using explicit args arrays. No shell parsing involved now — aligns with the repo convention that `exec()` is reserved for shell-only features.

File: source/operations/cleanupFiles.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+  await exec('rm -rf src/subgraphs', { cwd: projectFolder })
+
+  if (isFeatureSelected('demo', features)) {
+    const homeFolder = 'src/components/pageComponents/home'
+
+    await exec(`rm -rf ${homeFolder}/Examples/demos/subgraphs`, { cwd: projectFolder })
+    await exec(`rm ${homeFolder}/Examples/index.tsx`, { cwd: projectFolder })
+    await exec(`cp .install-files/home/Examples/index.tsx ${homeFolder}/Examples/index.tsx`, {
+      cwd: projectFolder,
+    })
```
[the remark this answers is not in the document]
Comment (fernandomg) [reply source/operations/cleanupFiles.ts]: Fixed in 52d4361. Same commit — all `exec()` calls in this file now use `execFile()` with args arrays, including the `homeFolder` template literal paths.

File: source/operations/exec.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+      stderr += data.toString()
+    })
+
+    child.on('close', (code) => {
+      if (code === 0) {
+        resolve()
+      } else {
+        const message = stderr.trim() || `Command failed with exit code ${code}`
+        reject(new Error(message))
+      }
```
[the remark this answers is not in the document]
Comment (fernandomg) [reply source/operations/exec.ts]: Fixed in 461ea7a. The `close` handler now receives `(code, signal)` and reports `Process killed by signal ${signal}` when `code` is null. Added exec.ts test suite covering signal kills, stderr errors, exit codes, and spawn errors.

File: source/operations/exec.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  return new Promise((resolve, reject) => {
+    const child = spawn(command, args, {
+      ...options,
+      stdio: ['ignore', 'ignore', 'pipe'],
+    })
+
+    let stderr = ''
+    child.stderr?.on('data', (data: Buffer) => {
+      stderr += data.toString()
+    })
```
[the remark this answers is not in the document]
Comment (fernandomg) [reply source/operations/exec.ts:13]: Acknowledged — in practice the commands we run (`rm`, `mkdir`, `cp`, `git`, `pnpm`) produce minimal stderr output. The longest-running commands (`pnpm i`, `git clone`) emit progress/warnings but not megabytes. We'll revisit if this becomes a real issue, but capping stderr now would add complexity for a theoretical scenario.

File: source/operations/exec.ts [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+      }
+    })
+
+    child.on('error', reject)
+  })
+}
+
+export async function exec(command: string, options: { cwd?: string } = {}): Promise<void> {
+  await run('/bin/sh', ['-c', command], { cwd: options.cwd })
+}
```
[the remark this answers is not in the document]
Comment (fernandomg) [reply source/operations/exec.ts:34]: This is a POSIX-only CLI — it shells out to `rm`, `cp`, `mkdir`, `git`, and `pnpm`, all assuming a Unix environment. Windows portability would require replacing all of those with Node fs equivalents, not just the shell path. Out of scope for this PR.

File: source/operations/cloneRepo.ts [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+
+export async function cloneRepo(projectName: string): Promise<void> {
+  const projectFolder = getProjectFolder(projectName)
+
+  await execFile('git', ['clone', '--depth', '1', '--no-checkout', repoUrl, projectName])
+  await execFile('git', ['fetch', '--tags'], { cwd: projectFolder })
+  // Shell required for $() command substitution
+  await exec('git checkout $(git describe --tags $(git rev-list --tags --max-count=1))', {
+    cwd: projectFolder,
+  })
```
[the remark this answers is not in the document]
Comment (fernandomg) [reply source/operations/cloneRepo.ts:23]: Same as the `/bin/sh` comment — the entire tool assumes POSIX. This is the only remaining `exec()` call and it legitimately needs shell substitution (`$()`). Making this shell-free would also require `execFile` to capture stdout, which it currently doesn't by design. Not worth the complexity for a non-goal.

File: architecture.md [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+
+## Key Abstractions
+
+### Feature Definitions (`source/constants/config.ts`)
+
+Single source of truth for all feature metadata. Every consumer reads from here:
+
+```ts
+featureDefinitions: Record<FeatureName, {
+  description: string   // --info output, --help text
```
[the remark this answers is not in the document]
Comment (fernandomg) [reply architecture.md]: Fixed in bcb2d44. Removed the `--help text` claim — the description field is only used in `--info` output.

File: source/components/steps/Install/Install.tsx [hunk 12]
```diff
@@ -19,55 +19,38 @@ interface Props {
...
-              selectedFeatures={selectedFeatures}
-              onCompletion={onCompletion}
-              projectFolder={projectFolder}
-            />
-          )}
-        </Script>
-      </Box>
+      {status === 'running' && <Text color={'whiteBright'}>Installing packages... Working...</Text>}
+      {status === 'done' && <Text color={'green'}>Installation complete. Done!</Text>}
+      {status === 'error' && <Text color={'red'}>Installation failed: {errorMessage}</Text>}
```
[the remark this answers is not in the document]
Comment (fernandomg) [reply source/components/steps/Install/Install.tsx]: Fixed in b1870e9. Added `onProgress` callback to `cloneRepo`, `installPackages`, and `cleanupFiles`. TUI components now render per-step progress with checkmarks (completed) and dimmed status (in-progress), matching the original ink-spawn stepper UX. Non-interactive path omits the callback.

File: source/cli.tsx [hunk 13]
```diff
@@ -1,6 +1,103 @@
...
+  Options
+    --name <string>          Project name (alphanumeric, underscores)
+    --mode <full|custom>     Installation mode
+    --features <list>        Comma-separated features (with --mode=custom):
+                               demo       Component demos and example pages
+                               subgraph   TheGraph subgraph integration (requires API key)
+                               typedoc    TypeDoc API documentation generation
+                               vocs       Vocs documentation site
+                               husky      Git hooks with Husky, lint-staged, commitlint
+    --non-interactive, --ni  Run without prompts (auto-enabled when not a TTY)
```
[the remark this answers is not in the document]
Comment (fernandomg) [reply source/cli.tsx:21]: Keeping as-is — this was raised before. The help text is short, unlikely to drift, and `--info` is the canonical discovery path for agents.

File: source/nonInteractive.ts [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+
+type ErrorResult = {
+  success: false
+  error: string
+}
+
+function fail(error: string): never {
+  const result: ErrorResult = { success: false, error }
+  console.log(JSON.stringify(result, null, 2))
+  process.exit(1)
```
[the remark this answers is not in the document]
Comment (fernandomg) [reply source/nonInteractive.ts]: Fixed in 106c6f0. Replaced `process.exit(1)` with `process.exitCode = 1` + throw in both `fail()` and the cli.tsx catch handler. The event loop now drains naturally, flushing stdout before exit. Ref: https://nodejs.org/api/process.html#processexitcode

File: source/cli.tsx [hunk 15]
```diff
@@ -1,6 +1,103 @@
...
+if (isNonInteractive) {
+  runNonInteractive({
+    name: cli.flags.name,
+    mode: cli.flags.mode,
+    features: cli.flags.features,
+  }).catch((error) => {
+    const message = error instanceof Error ? error.message : String(error)
+    console.log(JSON.stringify({ success: false, error: message }, null, 2))
+    process.exit(1)
+  })
```
[the remark this answers is not in the document]
Comment (fernandomg) [reply source/cli.tsx]: Fixed in 106c6f0. Same commit — the cli.tsx catch handler now uses `process.exitCode = 1` and guards against double-writing when `fail()` already handled the error.

File: source/cli.tsx [hunk 16]
```diff
@@ -1,6 +1,106 @@
...
+} else {
+  const run = async () => {
+    console.clear()
+    const { render } = await import('ink')
+    const { default: App } = await import('./app.js')
+
+    render(<App />)
+  }
+
+  run().catch(console.error)
```
[the remark this answers is not in the document]
Comment (fernandomg) [reply source/cli.tsx]: Fixed in c8dc7fb. Restructured cli.tsx routing as `if/else-if/else` — the `--info` path no longer calls `process.exit()`. No `process.exit()` calls remain anywhere in the source.

File: source/operations/cleanupFiles.ts [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+  await execFile('rm', ['-rf', 'src/subgraphs'], { cwd: projectFolder })
+
+  if (isFeatureSelected('demo', features)) {
+    const homeFolder = 'src/components/pageComponents/home'
+
+    await execFile('rm', ['-rf', `${homeFolder}/Examples/demos/subgraphs`], {
+      cwd: projectFolder,
+    })
+    await execFile('rm', [`${homeFolder}/Examples/index.tsx`], { cwd: projectFolder })
+    await execFile(
```
[the remark this answers is not in the document]
Comment (fernandomg) [reply source/operations/cleanupFiles.ts]: Fixed in d0e4592. Added `-f` to all single-file `rm` calls in `cleanupFiles.ts` for idempotent cleanup. Also added a guard test that ensures every `rm` command uses `-f` or `-rf`.

File: source/operations/cleanupFiles.ts [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+
+async function cleanupVocs(projectFolder: string): Promise<void> {
+  await execFile('rm', ['vocs.config.ts'], { cwd: projectFolder })
+  await execFile('rm', ['-rf', 'docs'], { cwd: projectFolder })
+}
+
+async function cleanupHusky(projectFolder: string): Promise<void> {
+  await execFile('rm', ['-rf', '.husky'], { cwd: projectFolder })
+  await execFile('rm', ['.lintstagedrc.mjs'], { cwd: projectFolder })
+  await execFile('rm', ['commitlint.config.js'], { cwd: projectFolder })
```
[the remark this answers is not in the document]
Comment (fernandomg) [reply source/operations/cleanupFiles.ts]: Fixed in d0e4592. Same commit — all single-file `rm` calls now use `-f`.

File: source/operations/cleanupFiles.ts [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+async function cleanupTypedoc(projectFolder: string): Promise<void> {
+  await execFile('rm', ['typedoc.json'], { cwd: projectFolder })
+}
+
+async function cleanupVocs(projectFolder: string): Promise<void> {
+  await execFile('rm', ['vocs.config.ts'], { cwd: projectFolder })
+  await execFile('rm', ['-rf', 'docs'], { cwd: projectFolder })
+}
```
[the remark this answers is not in the document]
Comment (fernandomg) [reply source/operations/cleanupFiles.ts:65]: Fixed in d0e4592. Same commit — all single-file `rm` calls now use `-f`.

File: source/operations/cleanupFiles.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (fernandomg) [reply source/operations/cleanupFiles.ts]: Fixed in d0e4592. Same commit — all single-file `rm` calls now use `-f`.

File: source/components/steps/FileCleanup.tsx [hunk 20]
```diff
@@ -16,242 +15,52 @@ interface Props {
...
- * @param onCompletion
- * @param installation
- * @param projectName
- */
-const FileCleanup: FC<Props> = ({ onCompletion, installationConfig, projectName }) => {
-  const { installationType, selectedFeatures } = installationConfig
-  const projectFolder = useMemo(() => getProjectFolder(projectName), [projectName])
-  const currentHomeFolder = `${projectFolder}/src/components/pageComponents/home`
+  const completedSteps = status === 'done' ? steps : steps.slice(0, -1)
+  const currentStep = status === 'running' ? steps.at(-1) : undefined
```
[the remark this answers is not in the document]
Comment (fernandomg) [reply source/components/steps/FileCleanup.tsx]: Fixed in 687f8d5. Extracted `deriveStepDisplay()` utility that computes `completedSteps`, `currentStep`, and `failedStep` from the steps array and status. All three TUI components now show the failed step with a `✗` red marker before the error message.

File: source/components/steps/CloneRepo/CloneRepo.tsx [hunk 21]
```diff
@@ -1,26 +1,57 @@
...
+      })
+      .catch((error: unknown) => {
+        setStatus('error')
+        setErrorMessage(error instanceof Error ? error.message : String(error))
+      })
+  }, [projectName, onCompletion, handleProgress])
+
+  const completedSteps = status === 'done' ? steps : steps.slice(0, -1)
+  const currentStep = status === 'running' ? steps.at(-1) : undefined
+
```
[the remark this answers is not in the document]
Comment (fernandomg) [reply source/components/steps/CloneRepo/CloneRepo.tsx]: Fixed in 687f8d5. Same commit — shared `deriveStepDisplay()` handles the state derivation for all three components.

File: source/components/steps/Install/Install.tsx [hunk 22]
```diff
@@ -19,55 +19,59 @@ interface Props {
...
+      })
+      .catch((error: unknown) => {
+        setStatus('error')
+        setErrorMessage(error instanceof Error ? error.message : String(error))
+      })
+  }, [projectFolder, installationType, selectedFeatures, onCompletion, handleProgress])
+
+  const completedSteps = status === 'done' ? steps : steps.slice(0, -1)
+  const currentStep = status === 'running' ? steps.at(-1) : undefined
+
```
[the remark this answers is not in the document]
Comment (fernandomg) [reply source/components/steps/Install/Install.tsx]: Fixed in 687f8d5. Same commit.

File: AGENTS.md [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+- **Trailing commas:** none (Biome default)
+- **Indent:** spaces, width 2
+- **Imports:** explicit `.js` extensions (ESM, `"type": "module"`)
+
+## Working Rules
+
+- Use **pnpm** only (never npm or yarn)
+- Treat `dist/` as build output — never edit directly
+- User input (`projectName`) must never be interpolated into shell command strings — use `execFile` (args array) instead
+- `source/constants/config.ts` is the single source of truth for feature metadata — all consumers read from it
```
[the remark this answers is not in the document]
Comment (fernandomg) [reply AGENTS.md]: Fixed in 66468bd. Updated to: "all programmatic consumers read from it (CLI `--help` text maintains its own copy)".

File: architecture.md [hunk 24]
```diff
@@ -0,0 +1,191 @@
...
+      createEnvFile.test.ts
+      installPackages.test.ts
+      cleanupFiles.test.ts
+```
+
+## Key Abstractions
+
+### Feature Definitions (`source/constants/config.ts`)
+
+Single source of truth for all feature metadata. Every consumer reads from here:
```
[the remark this answers is not in the document]
Comment (fernandomg) [reply architecture.md]: Fixed in 66468bd. Same commit — clarified which surfaces are derived from `featureDefinitions` and noted the `--help` exception.

File: source/components/steps/Install/Install.tsx [hunk 25]
```diff
@@ -19,55 +19,59 @@ interface Props {
 const Install: FC<Props> = ({ projectName, onCompletion, installationConfig }) => {
   const { installationType, selectedFeatures } = installationConfig
   const projectFolder = useMemo(() => getProjectFolder(projectName), [projectName])
+  const [steps, setSteps] = useState<string[]>([])
+  const [status, setStatus] = useState<'running' | 'done' | 'error'>('running')
+  const [errorMessage, setErrorMessage] = useState('')
+
   const title = installationType
     ? installationType[0]?.toUpperCase() + installationType.slice(1)
     : ''
```
[the remark this answers is not in the document]
Comment (fernandomg) [reply source/components/steps/Install/Install.tsx]: Fixed in 4f33846. Changed the ternary default from `''` to `'Full'` directly and removed the dead `?? 'Full'` from the template literal.

File: source/operations/cleanupFiles.ts [hunk 26]
```diff
@@ -0,0 +1,110 @@
+import { readFileSync, writeFileSync } from 'node:fs'
+import { copyFile, mkdir, rm } from 'node:fs/promises'
```
[the remark this answers is not in the document]
Comment (fernandomg) [reply source/operations/cleanupFiles.ts:2]: The architecture.md already documents this exception: "`patchPackageJson` uses sync `node:fs`". Updated the PR description to match — it now says "sync `node:fs` for `patchPackageJson` and `existsSync` checks". The sync calls are intentional for these small blocking reads.

File: source/utils/utils.ts [hunk 27]
```diff
@@ -1,57 +1,63 @@
...
+  if (mode === 'full') {
+    return Object.values(featureDefinitions).flatMap((def) => def.postInstall ?? [])
+  }
+
+  return selectedFeatures.flatMap((name) => featureDefinitions[name]?.postInstall ?? [])
+}
+
+export function projectDirectoryExists(projectName: string): boolean {
+  return existsSync(getProjectFolder(projectName))
+}
```
[the remark this answers is not in the document]
Comment (fernandomg) [reply source/utils/utils.ts:47]: Same as above — `existsSync` is a small blocking check in the validation path before any async operations run. Updated PR description to acknowledge sync `node:fs` usage for these cases.

[the remark this answers is not in the document]
Comment (fernandomg) [reply source/utils/utils.ts:47]: In practice this checks whether a project _can_ be created at that path — a file collision is just as much a conflict as a directory one. Renaming to `projectPathExists` would be more precise but also more generic than the intent: we are specifically guarding the project directory creation step. The current name + error message are clear enough for the use case. Not worth the churn.
